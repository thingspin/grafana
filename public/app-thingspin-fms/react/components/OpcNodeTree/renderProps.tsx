import React, { MouseEvent } from 'react';

const LEVEL_SPACE = 20;

type ToggleIconParams = {
  on: boolean;
  onClick: (event: MouseEvent<HTMLDivElement>) => void;
};

const ToggleIcon = ({ on, onClick }: ToggleIconParams): JSX.Element => (
  <div className="rc-opc-node-toggle-icon" role="button" aria-label="Toggle" onClick={onClick}>
    <i className={on ? "fa fa-caret-down" : "fa fa-caret-up"}></i>
  </div>
);

export interface NodeTreeItem {
  hasNodes?: boolean;
  isOpen?: boolean;
  level?: number;
  active?: boolean;
  key: string;
  label: string | JSX.Element;
  onClick?: (event: MouseEvent<HTMLLIElement>) => void;
  onClickOnBtn?: (event: MouseEvent<HTMLDivElement>) => void;
  onAddBtn?: (event: MouseEvent<HTMLButtonElement>) => void;
  [name: string]: any;
}

export type NodeTreeChildren = (
  props: {
    search: Function;
    items: NodeTreeItem[];
  }
) => JSX.Element;

type RenderItem = (props: NodeTreeItem) => JSX.Element;

const renderItem: RenderItem = ({
  hasNodes = false,
  isOpen = false,
  level = 0,
  onClick,
  onClickOnBtn,
  onAddBtn,
  active,
  key,
  label = 'unknown',
}): JSX.Element => (
    <li aria-pressed={active} key={key} onClick={onClick}
      style={{
        padding: ` 12px  0  12px ${level * LEVEL_SPACE}px`,
        color: '#1e1e1e',
        background: active ? '#e3f2fd' : 'none',
      }}
    >
      {(<div className="rc-opc-node-tree"
          style={{
            left: `${level * LEVEL_SPACE}px`,
          }}
        >
          <ToggleIcon on={isOpen} onClick={onClickOnBtn} />

          <div className="rc-opc-node-icon"><i className="fa fa-folder"></i></div>

          {label}

          <button className="rc-opc-node-button" onClick={onAddBtn}><i className="fa fa-plus-circle"></i></button>
        </div>
      )}
    </li>
  );

export const defaultChildren: NodeTreeChildren = ({ search, items }): JSX.Element => {
  const onSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = e.target;
    search(value);
  };

  return (
    <>
      <input type="search" placeholder="Type and search" aria-label="Type and search"
        className="rc-opc-tree-search" onChange={onSearch} />

      <ul style={{ listStyleType: 'none', paddingLeft: 0, overflow: "auto", }}>
        {items.map(renderItem)}
      </ul>
    </>
  );
};
