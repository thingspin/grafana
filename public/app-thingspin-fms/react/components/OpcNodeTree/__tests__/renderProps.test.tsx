import { mount } from 'enzyme';

import { defaultChildren } from '../renderProps';

const search = () => console.log('search');

describe('defaultChildren', () => {
  it('should render without the toggle icon', () => {
    const wrapper = mount(
      defaultChildren({ search, items: [{ hasNodes: false, label: 'foo', key: 'key' }] })
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ToggleIcon').exists()).toEqual(false);
  });

  it('should render the toggle icon with "on" equals to true', () => {
    const wrapper = mount(
      defaultChildren({
        search,
        items: [
          {
            hasNodes: true,
            isOpen: true,
            level: 1,
            label: 'foo',
            key: 'foo',
          },
          {
            hasNodes: false,
            isOpen: false,
            level: 2,
            label: 'bar',
            key: 'bar',
          },
          {
            hasNodes: false,
            isOpen: false,
            level: 2,
            label: 'zoo',
            key: 'zoo',
          },
        ],
      })
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ToggleIcon').prop('on')).toEqual(true);
  });
  it('should render the toggle icon with "on" equals to false', () => {
    const wrapper = mount(
      defaultChildren({
        search,
        items: [
          {
            hasNodes: true,
            isOpen: false,
            level: 1,
            label: 'foo',
            key: 'key',
          },
        ],
      })
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('ToggleIcon').prop('on')).toEqual(false);
  });
});
