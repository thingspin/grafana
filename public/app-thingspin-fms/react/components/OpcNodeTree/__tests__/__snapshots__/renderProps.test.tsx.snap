// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultChildren should render the toggle icon with "on" equals to false 1`] = `
Array [
  <input
    aria-label="Type and search"
    onChange={[Function]}
    placeholder="Type and search"
    style={
      Object {
        "border": "none",
        "padding": "1rem 2rem",
        "width": "100%",
      }
    }
    type="search"
  />,
  <ul
    style={
      Object {
        "borderTop": "1px solid #ccc",
        "listStyleType": "none",
        "paddingLeft": 0,
      }
    }
  >
    <li
      key="key"
      role="button"
      style={
        Object {
          "background": "none",
          "borderBottom": "1px solid #ccc",
          "color": "#333",
          "cursor": "pointer",
          "padding": " .75rem  1rem  .75rem 4.75rem",
        }
      }
    >
      <div
        style={
          Object {
            "height": "2rem",
            "left": "3.25rem",
            "position": "absolute",
            "width": "2rem",
          }
        }
      >
        <ToggleIcon
          on={false}
        >
          <div
            aria-label="Toggle"
            role="img"
          >
            +
          </div>
        </ToggleIcon>
      </div>
      foo
    </li>
  </ul>,
]
`;

exports[`defaultChildren should render the toggle icon with "on" equals to true 1`] = `
Array [
  <input
    aria-label="Type and search"
    onChange={[Function]}
    placeholder="Type and search"
    style={
      Object {
        "border": "none",
        "padding": "1rem 2rem",
        "width": "100%",
      }
    }
    type="search"
  />,
  <ul
    style={
      Object {
        "borderTop": "1px solid #ccc",
        "listStyleType": "none",
        "paddingLeft": 0,
      }
    }
  >
    <li
      key="foo"
      role="button"
      style={
        Object {
          "background": "none",
          "borderBottom": "1px solid #ccc",
          "color": "#333",
          "cursor": "pointer",
          "padding": " .75rem  1rem  .75rem 4.75rem",
        }
      }
    >
      <div
        style={
          Object {
            "height": "2rem",
            "left": "3.25rem",
            "position": "absolute",
            "width": "2rem",
          }
        }
      >
        <ToggleIcon
          on={true}
        >
          <div
            aria-label="Toggle"
            role="img"
          >
            -
          </div>
        </ToggleIcon>
      </div>
      foo
    </li>
    <li
      key="bar"
      role="button"
      style={
        Object {
          "background": "none",
          "borderBottom": "1px solid #ccc",
          "color": "#333",
          "cursor": "pointer",
          "padding": " .75rem  1rem  .75rem 6rem",
        }
      }
    >
      bar
    </li>
    <li
      key="zoo"
      role="button"
      style={
        Object {
          "background": "none",
          "borderBottom": "1px solid #ccc",
          "color": "#333",
          "cursor": "pointer",
          "padding": " .75rem  1rem  .75rem 6rem",
        }
      }
    >
      zoo
    </li>
  </ul>,
]
`;

exports[`defaultChildren should render without the toggle icon 1`] = `
Array [
  <input
    aria-label="Type and search"
    onChange={[Function]}
    placeholder="Type and search"
    style={
      Object {
        "border": "none",
        "padding": "1rem 2rem",
        "width": "100%",
      }
    }
    type="search"
  />,
  <ul
    style={
      Object {
        "borderTop": "1px solid #ccc",
        "listStyleType": "none",
        "paddingLeft": 0,
      }
    }
  >
    <li
      key="key"
      role="button"
      style={
        Object {
          "background": "none",
          "borderBottom": "1px solid #ccc",
          "color": "#333",
          "cursor": "pointer",
          "padding": " .75rem  1rem  .75rem 3.5rem",
        }
      }
    >
      foo
    </li>
  </ul>,
]
`;
